### PR TITLE
Added ELF Loading from USB/HDD and Mira ELF Updating for HDD etc 

### PR DESCRIPTION
**Pull request by LightningMods**

### CODE REVIEW

File: loader/src/LoaderEntry.cpp [hunk 1]
```diff
@@ -57,10 +58,54 @@ int(*sceNetListen)(int, int) = nullptr;
...
+
+
+char* usbpath()
+{
+	int usb;
+	static char usbbuf[100];
+	usbbuf[0] = '\0';
+	for (int x = 0; x <= 7; x++)
+	{
+		snprintf(usbbuf, 100, "/mnt/usb%i/.dirtest", x);
```
Comment (kiwidoggie) [inline loader/src/LoaderEntry.cpp]: Change to sizeof(usbbuf), we want as little constants as possible

File: loader/src/LoaderEntry.cpp [hunk 2]
```diff
@@ -57,10 +58,54 @@ int(*sceNetListen)(int, int) = nullptr;
...
+		usb = sceKernelOpen(usbbuf, 0x0001 | 0x0200 | 0x0400, 0777);
```
Comment (kiwidoggie) [inline loader/src/LoaderEntry.cpp]: Use the proper flags, we don't want to have to guess what 0x0001 | 0x0200 etc are

File: loader/src/LoaderEntry.cpp [hunk 3]
```diff
@@ -121,6 +166,109 @@ void mira_escape(struct thread* td, void* uap)
...
+   
+
+
+    MD5_Init (&mdContext2);
+    while ((bytes2 = fread (data2, 1, 1024, hdd)) != 0)
+        MD5_Update (&mdContext2, data2, bytes2);
+    MD5_Final (c2,&mdContext2);
+
+
+  printf("MD5 HASH OF USB ELF IS %i%i%i%i%i%i%i%i%i%i%i%i%i%i%i\n", c[1], c[2], c[3], c[4], c[5], c[6], c[7], c[8], c[9], c[10], c[11], c[12], c[13], c[14], c[15]);
```
Comment (kiwidoggie) [inline loader/src/LoaderEntry.cpp]: Use iteration here

File: loader/src/LoaderEntry.cpp [hunk 4]
```diff
@@ -121,6 +166,109 @@ void mira_escape(struct thread* td, void* uap)
...
+	 unsigned char c[16];
+	 unsigned char c2[16];
+    int i;
+    FILE *usb = fopen (usbfile, "rb");
+    FILE *hdd = fopen ("/user/MiraLoader.elf", "rb");
+    MD5_CTX mdContext;
+
+    MD5_CTX mdContext2;
+    int bytes2 = 0;
+    unsigned char data2[1024];
```
Comment (kiwidoggie) [inline loader/src/LoaderEntry.cpp:181]: Hash the entire executable, not just the first kb

File: loader/src/LoaderEntry.cpp [hunk 5]
```diff
@@ -121,6 +166,109 @@ void mira_escape(struct thread* td, void* uap)
 	printf("[-] mira_escape\n");
 }
 
+#define DIFFERENT_HASH 1
+#define SAME_HASH 0
+
+int MD5_hash_compare(const char *usbfile)
+{
+	 unsigned char c[16];
```
Comment (kiwidoggie) [inline loader/src/LoaderEntry.cpp]: constant here, like MD5_HASH_LENGTH or something

File: loader/src/LoaderEntry.cpp [hunk 6]
```diff
@@ -121,6 +166,109 @@ void mira_escape(struct thread* td, void* uap)
...
+
+    fclose (usb);
+    fclose (hdd);
+
+return SAME_HASH;
+}
+
+     int ftruncate(int fd, off_t length)
+     {
+     	return syscall(480, fd, length);
```
Comment (kiwidoggie) [inline loader/src/LoaderEntry.cpp:218]: use syscall2

File: loader/src/LoaderEntry.cpp [hunk 7]
```diff
@@ -121,6 +166,109 @@ void mira_escape(struct thread* td, void* uap)
...
+     }
+
+
+    int munmap(void *addr,	size_t len)
+    {
+    	return syscall(73, addr, len);
```
Comment (kiwidoggie) [inline loader/src/LoaderEntry.cpp:223]: use syscall2

File: loader/src/LoaderEntry.cpp [hunk 8]
```diff
@@ -121,6 +166,109 @@ void mira_escape(struct thread* td, void* uap)
...
+    /* SOURCE */
+    sfd = sceKernelOpen(sourcefile, O_RDONLY, 0x000);
+    filesize = sceKernelLseek(sfd, 0, SEEK_END);
+
+    if(filesize < 0) return -1;
+
+    src = _mmap(NULL, filesize, PROT_READ, MAP_PRIVATE, sfd, 0);
+
+    /* DESTINATION */
+    dfd = sceKernelOpen("/user/MiraLoader.elf", O_RDWR | O_CREAT | O_TRUNC, 0666);
```
Comment (kiwidoggie) [inline loader/src/LoaderEntry.cpp]: Are these permissions okay? Should it be 0777?

File: loader/src/LoaderEntry.cpp [hunk 9]
```diff
@@ -121,6 +166,109 @@ void mira_escape(struct thread* td, void* uap)
...
+
+    ftruncate(dfd, filesize);
```
Comment (kiwidoggie) [inline loader/src/LoaderEntry.cpp:244]: Is this needed if you are passing O_TRUNC?

File: loader/src/LoaderEntry.cpp [hunk 10]
```diff
@@ -215,6 +396,106 @@ extern "C" void* mira_entry(void* args)
...
+
+if (strlen(usbpath()) != 0 ||  hddfile > 0)
+{
+ char filebuffer[200] = { 0 };
+
+
+
+
+snprintf(filebuffer, 200, "%s/MiraLoader.elf", usbpath());
+int filefd = sceKernelOpen(filebuffer, 0x000, 0x000);
```
Comment (kiwidoggie) [inline loader/src/LoaderEntry.cpp]: use proper flags

File: loader/src/LoaderEntry.cpp [hunk 11]
```diff
@@ -215,6 +396,106 @@ extern "C" void* mira_entry(void* args)
 	// Close the client and server socket connections
 	sceNetSocketClose(clientSocket);
 	sceNetSocketClose(serverSocket);
+}
+
+int hddfile = sceKernelOpen("/user/MiraLoader.elf", 0x0000, 0x0000);
```
Comment (kiwidoggie) [inline loader/src/LoaderEntry.cpp]: use proper flags

File: loader/src/LoaderEntry.cpp [hunk 12]
```diff
@@ -215,6 +396,106 @@ extern "C" void* mira_entry(void* args)
 	// Close the client and server socket connections
 	sceNetSocketClose(clientSocket);
 	sceNetSocketClose(serverSocket);
+}
+
+int hddfile = sceKernelOpen("/user/MiraLoader.elf", 0x0000, 0x0000); 
+
+if (strlen(usbpath()) != 0 ||  hddfile > 0)
+{
+ char filebuffer[200] = { 0 };
```
Comment (kiwidoggie) [inline loader/src/LoaderEntry.cpp]: use the max path (I think it's MAX_PATH, or _MAX_PATH len 260?)

File: loader/src/LoaderEntry.cpp [hunk 13]
```diff
@@ -215,6 +396,106 @@ extern "C" void* mira_entry(void* args)
...
+if(filefd > 0)
+{
+	WriteNotificationLog("Found USB");
+	printf("USB at %s\n", filebuffer);
+	//WriteNotificationLog(filebuffer);
+}
+
+
+
+if (filefd > 0)
```
Comment (kiwidoggie) [inline loader/src/LoaderEntry.cpp]: invert this check to prevent indention

File: loader/src/LoaderEntry.cpp [hunk 14]
```diff
@@ -215,6 +396,106 @@ extern "C" void* mira_entry(void* args)
...
+{
+
+	if (sceKernelOpen("/user/MiraLoader.elf", 0x0000, 0x0000) > 0)
```
Comment (kiwidoggie) [inline loader/src/LoaderEntry.cpp]: use proper flags

File: loader/src/LoaderEntry.cpp [hunk 15]
```diff
@@ -215,6 +396,106 @@ extern "C" void* mira_entry(void* args)
...
+else
+{
+	WriteNotificationLog("Loading from HDD no USB");
+	printf("Loading from HDD no USB\n");
+	filefd = hddfile;
+}
+	
+
+
+    currentSize = sceKernelLseek(filefd, 0, 2);
```
Comment (kiwidoggie) [inline loader/src/LoaderEntry.cpp]: use proper flags

File: loader/src/Utils/Syscall.s [hunk 16]
```diff
@@ -1,13 +1,15 @@
 .intel_syntax noprefix
 .text
 
-.global syscall1, syscall2, syscall3, syscall4, syscall5, syscall6, _mmap
+.global syscall, syscall1, syscall2, syscall3, syscall4, syscall5, syscall6, _mmap
 
 syscall:
-    mov rax,rdi
+    mov rax,0
```
Comment (kiwidoggie) [inline loader/src/Utils/Syscall.s:7]: shouldn't need this, but good for additional coverage

**Review by kiwidoggie — CHANGES_REQUESTED** [review]
Good start

File: loader/src/LoaderEntry.cpp (hunk 4 above)
Comment (LightningMods) [reply loader/src/LoaderEntry.cpp:181]: it does hash the whole elf in the while loop 1024 at a time

File: loader/src/LoaderEntry.cpp (hunk 8 above)
Comment (LightningMods) [reply loader/src/LoaderEntry.cpp]: seems to make it fine , illl change it tho

File: loader/src/LoaderEntry.cpp (hunk 9 above)
Comment (LightningMods) [reply loader/src/LoaderEntry.cpp:244]: O_TRUNC 0s it and ftruncate makes the file the correct size iirc

**Comment by LightningMods** [comment]
rebasing into 1 commit

**Pull request closed without merging**